lib/ProductOpener/Products.pm

```diff
@@ -1024,52 +1025,68 @@ sub retrieve_product_rev ($product_id, $rev, $include_deleted = 0) {
 
 sub change_product_server_or_code ($product_ref, $new_code, $errors_ref) {
 
-	# Currently only called by admins, can cause issues because of bug #677
+	# Currently only called by admins and moderators
 
 	my $code = $product_ref->{code};
-	my $new_server = "";
-	my $new_data_root = $data_root;
 
 	if ($new_code =~ /^([a-z]+)$/) {
-		$new_server = $1;
-		if (    (defined $options{other_servers})
-			and (defined $options{other_servers}{$new_server})
-			and ($options{other_servers}{$new_server}{data_root} ne $data_root))
-		{
-			$new_code = $code;
-			$new_data_root = $options{other_servers}{$new_server}{data_root};
+		my $new_server = $1;
+		if (defined $options{flavors_product_types}{$new_server}) {
+			change_product_type($product_ref, $options{flavors_product_types}{$new_server}, $errors_ref);
 		}
 	}
 
-	$new_code = normalize_code($new_code);
-	if (not is_valid_code($new_code)) {
-		push @$errors_ref, lang("invalid_barcode");
-	}
 	else {
-		# check that the new code is available
-		if (-e "$new_data_root/products/" . product_path_from_id($new_code)) {
-			push @{$errors_ref}, lang("error_new_code_already_exists");
-			$log->warn(
-				"cannot change product code, because the new code already exists",
-				{code => $code, new_code => $new_code, new_server => $new_server}
-			) if $log->is_warn();
+		$new_code = normalize_code($new_code);
+		if (not is_valid_code($new_code)) {
+			push @$errors_ref, lang("invalid_barcode");
 		}
 		else {
-			$product_ref->{old_code} = $code;
-			$code = $new_code;
-			$product_ref->{code} = $code;
-			if ($new_server ne '') {
-				$product_ref->{new_server} = $new_server;
-			}
-			$log->info("changing code",
-				{old_code => $product_ref->{old_code}, code => $code, new_server => $new_server})
-				if $log->is_info();
+			# check that the new code is available
+			if (-e "$data_root/products/" . product_path_from_id($new_code) . "/product.sto") {
+				push @{$errors_ref}, lang("error_new_code_already_exists");
+				$log->warn("cannot change product code, because the new code already exists",
+					{code => $code, new_code => $new_code})
+					if $log->is_warn();
+			}
+			else {
+				$product_ref->{old_code} = $code;
+				$code = $new_code;
+				$product_ref->{code} = $code;
+				$log->info("changing code", {old_code => $product_ref->{old_code}, code => $code})
+					if $log->is_info();
+			}
 		}
 	}
 
 	return;
 }
 
+sub change_product_type ($product_ref, $new_product_type, $errors_ref) {
```

Comment:
Adding a POD would be interesting.

Reply:
Good idea. I will also remove the feature of being able to change the product type by putting "obf" in the barcode field.